runtime/src/accounts_db.rs

```diff
@@ -1632,27 +1637,77 @@ impl AccountsDB {
         hasher.result()
     }
 
-    fn accumulate_account_hashes(mut hashes: Vec<(Pubkey, Hash)>) -> Hash {
-        let mut sort = Measure::start("sort");
+    fn accumulate_account_hashes(hashes: Vec<(Pubkey, Hash, u64)>) -> Hash {
+        let (hash, ..) = Self::do_accumulate_account_hashes_and_capitalization(hashes, false);
+        hash
+    }
+
+    fn accumulate_account_hashes_and_capitalization(
+        hashes: Vec<(Pubkey, Hash, u64)>,
+    ) -> (Hash, u64) {
+        let (hash, cap) = Self::do_accumulate_account_hashes_and_capitalization(hashes, true);
+        (hash, cap.unwrap())
+    }
+
+    fn do_accumulate_account_hashes_and_capitalization(
+        mut hashes: Vec<(Pubkey, Hash, u64)>,
+        calculate_cap: bool,
+    ) -> (Hash, Option<u64>) {
+        let mut sort_time = Measure::start("sort");
         hashes.par_sort_by(|a, b| a.0.cmp(&b.0));
-        sort.stop();
-        let mut hash_time = Measure::start("hash");
+        sort_time.stop();
 
-        let fanout = 16;
+        let mut sum_time = Measure::start("cap");
+        let cap = if calculate_cap {
+            Some(Self::checked_sum_for_capitalization(
+                hashes.iter().map(|(_, _, lamports)| *lamports),
+            ))
+        } else {
+            None
+        };
+        sum_time.stop();
 
+        let mut hash_time = Measure::start("hash");
+        let fanout = 16;
         let res = Self::compute_merkle_root(hashes, fanout);
-
         hash_time.stop();
-        debug!("{} {}", sort, hash_time);
 
-        res
+        debug!("{} {} {}", sort_time, hash_time, sum_time);
+
+        (res, cap)
+    }
+
+    pub fn checked_sum_for_capitalization<T: Iterator<Item = u64>>(balances: T) -> u64 {
+        balances
+            .map(|b| b as u128)
+            .sum::<u128>()
+            .try_into()
+            .expect("overflow is detected while summing capitalization")
+    }
+
+    pub fn account_balance_for_capitalization(
+        lamports: u64,
+        owner: &Pubkey,
+        executable: bool,
+    ) -> u64 {
+        let is_specially_retained = (solana_sdk::native_loader::check_id(owner) && executable)
+            || solana_sdk::sysvar::check_id(owner);
+
+        if is_specially_retained {
+            // specially retained accounts are ensured to exist by
+            // always having a balance of 1 lamports, which is
+            // outside the capitalization calculation.
+            lamports - 1
```

Comment:
Just return 0?

Reply:
Or is it because anyone could transfer to sysvars? If that's the case, I think if we can reword the comment to say something like "specially retained accounts always have an initial 1 lamport balance, but could be modified by transfers which increase the balance but don't affect the capitalization"

Reply:
Comment clarified: 22c1f2c9ee90410e8ffef6960dcee15de28258a6